### PR TITLE
stop puma from sending INT on SIGHUP which we use for logrotate

### PR DESCRIPTION
**Pull request by grosser**
https://github.com/rails/rails/issues/32211

/cc @ragurney @jonmoter

### CODE REVIEW

File: lib/samson/logging.rb [hunk 1]
```diff
@@ -27,3 +27,23 @@
   config.logger = Syslog::Logger.new('samson')
   config.lograge.formatter = Lograge::Formatters::Logstash.new
 end
+
+# puma kills itself with INT when it receives a SIGHUP see https://github.com/puma/puma/blob/master/docs/signals.md
+# If we have a file based logger then we need to reopen that file to log to the new logfile after log rotation.
+# The rails logger does not know what file it opened, so we have to help it https://github.com/rails/rails/issues/32211
+# If we do not have a file-based logger we still do not want to kill the process, but log that we got the signal.
+# We need to use self-pipe since we cannot reopen logs in an interrupt.
+# This will break stdout_redirect feature of puma, which I hope nobody uses.
```
Comment (jonmoter) [inline lib/samson/logging.rb]: Nice comment!  <3

File: lib/samson/logging.rb [hunk 2]
```diff
@@ -27,3 +27,23 @@
...
+if ENV["SERVER_MODE"]
+  read, write = IO.pipe
+  Signal.trap(:SIGHUP) { write.puts }
```
Comment (jonmoter) [inline lib/samson/logging.rb]: Doesn't this swallow SIGHUP entirely?  Don't we need to propagate the signal somehow?

File: lib/samson/logging.rb [hunk 3]
```diff
@@ -27,3 +27,23 @@
...
+  Thread.new do
+    loop do
+      read.gets
+      Rails.logger.info "Received SIGHUP ... reopening logs"
+      dev = Rails.logger.instance_variable_get(:@logdev)&.dev
+      path = dev.path if dev&.is_a?(File)
+      Rails.logger.reopen(path)
```
Comment (jonmoter) [inline lib/samson/logging.rb]: We want to call `Rails.logger.reopen(nil)` if `dev` isn't a file?

File: lib/samson/logging.rb (hunk 2 above)
Comment (grosser) [reply lib/samson/logging.rb]: swallowing is exactly what we want since puma translates HUP -> INT

File: lib/samson/logging.rb (hunk 3 above)
Comment (grosser) [reply lib/samson/logging.rb]: Calling with nil is a noop for all the other loggers, so not calling it is fine too ... maybe more obvious what is going on by not calling it

**Comment by grosser** [comment]
reworked to make our hack more obvious since we don't need all this file reopen logic anyway ...

looks good ?

File: lib/samson/logging.rb [hunk 4]
```diff
@@ -26,4 +26,26 @@
...
+  # This will break stdout_redirect feature of puma since we cannot support both
+  read, write = IO.pipe
+  Signal.trap(:SIGHUP) { write.puts }
+  Thread.new do
+    loop do
+      read.gets
+      Rails.logger.info "Received SIGHUP ... reopening logs"
+      # Rails.logger does not know what file it opened, so we help it https://github.com/rails/rails/issues/32211
+      dev = Rails.logger.instance_variable_get(:@logdev)&.dev
+      Rails.logger.reopen(dev.path)
```
Comment (jonmoter) [inline lib/samson/logging.rb]: Isn't it possible for `dev` to be `nil` here, so you'd be calling `nil.path`?

Comment (grosser) [reply lib/samson/logging.rb]: no since this must be a file logger at that point ... I'll remove the &. too that was ;eftover from previous approach

**Pull request merged**